music/src/piano_genie/model.ts

```diff
@@ -0,0 +1,295 @@
+/**
+ * Core implementation for Piano Genie model.
+ *
+ * @license
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Imports
+ */
+import * as tf from '@tensorflow/tfjs-core';
+
+/**
+ * Constants.
+ */
+const DATA_TIME_QUANTIZE_RATE = 31.25;
+const DATA_MAX_DISCRETE_TIMES = 32;
+const RNN_NLAYERS = 2;
+const RNN_NUNITS = 128;
+const NBUTTONS = 8;
+const NPIANOKEYS = 88;
+
+/**
+ * A type for keeping track of LSTM state activations.
+ *
+ * @param c The c parameters of the LSTM.
+ * @param h The h parameters of the LSTM
+ */
+type LSTMState = { c: tf.Tensor2D[], h: tf.Tensor2D[] };
+
+/**
+ * Creates empty LSTM state.
+ */
+function createZeroState() {
```

Comment:
Seems like this should be a class instead, and `createZeroState` should be the code in the constructor, and `dispose()` a function on the class. See https://github.com/tensorflow/magenta-js/blob/master/music/src/music_vae/model.ts#L37 for an example

Reply:
Agreed

Reply:
Ah I now remember why I did this previously. The tf.tidy() function doesn't work with container classes, only types. Going to keep this as is.

Reply:
shouldn't it handle any tensors that are created inside of the tidy regardless of where they are created?

Reply:
The problem is I want to *return* the LSTMState from the tf.tidy call. TFJS is unable to return arbitrary classes from calls to `tf.tidy`. I could concat all the state into a tf.Tensor2D but this would require an unpack and a pack on every call to `next`, which may harm latency.